### ISSUE TITLE
`Lowcoder: Visual Programming + Large Language Model`- AI for Low-Code for AI and related paper

### ISSUE THREAD

**Issue by xp1632** [xp1632/Aalen_working_log#5]
###  paper `AI for low-code for AI`

**Research Contribution**
- this paper argues that visual programming component compensates for the unambiguity caused by the natural language programming component
-  while the `discoverability` of unknown API in visual programming is improved by natural language component

---
**Tech things**

- After thoroughly reading this paper, we find out the `LowCoder(Natural Language)` part is functioned as `one natural language query`   <-> `one sk-learn operator`

- The `formulation of ML Pipeline` is still manual by the participants

- the formed pipeline displayed as `Blockly` blocks will then be translated into concatenation of code strings from each block
-  then  `Lale` recognize the final long code string and turn into an actual ml pipeline, code in `lowcoder_artifacts\tool\static\js\block_to_python_autogen.js` in [source code](https://zenodo.org/records/7601206)
![image](https://github.com/user-attachments/assets/1816875c-ebc4-4e0e-b85d-0293d9a4ae28)

---

**Other take-away**

- **Dataset**
- This paper used `proxy dataset` which is `nl-invocation code` pair from Kaggle competition and did some filtering

-  Training with `Hybrid Operator Invocation` where the Operation Name, provided parameter value, unprovided parameter value as placeholder achieve better performance, because it gives llm space to leave the unknown parameter as placeholder instead of hallucination 
 
- **Model**
- `fine-tuned CodeT5` model achieved better accuracy than `train from scratch` and `few-shot-prompted CodeGen`
- the reason that they don't fine-tuning CodeGen is because it's too large and expensive to train
- Despite the model hasan  accuracy of 73.57% to only generate `Operator Name`, to generate the whole `Operator invocation code` the accuracy is 41.27%
![image](https://github.com/user-attachments/assets/24e4f7ce-c4e1-494e-94ac-b708526a6b31)

-  
- **User Study**
- their user study deliberately includes `novice` and `non-novice` to test the interface for people with various level of ml-expertise
- they land their main contribution in the `discoverability` of `unknown API` for novice users who `knows what they want but don't know how`, the comparison groups of `NL interface` is a `simple keyword search`

- another thing they shows in the user study is that their tool allows user to iteratively refine the pipeline

----

- in its related paper: 

1.1 **Expectation vs. Experience: Evaluating the Usability of Code**
Generation Tools Powered by Large Language Models
https://dl.acm.org/doi/pdf/10.1145/3491101.3519665

- mentioned the generated code by llm model is hard to understand for users 
- also decomposition of complex task help improve the performance of llm

- highlight different parts of code and generate test case help user understand and debug the code

---

1.2 **CodeGen: An Open Large Language Model for Code with Multi-Turn Program Synthesis**

- 2022, llm model for program synthesis 
- https://arxiv.org/abs/2203.13474

---

1.4 **Pipeline Combinators for Gradual AutoML**

https://proceedings.neurips.cc/paper/2021/file/a3b36cb25e2e0b93b5f334ffb4e4064e-Paper.pdf

- github slides and code: https://github.com/IBM/lale/blob/master/talks/2019-1105-lale.pdf

- or Gradual AutoML using Lale https://dl.acm.org/doi/10.1145/3534678.3542630
- introduce `Lale` 
,  a sklearn-compatible library that offers compositional automated machine learning for data scientist and the **ml pipeline can be visualized as graphical pipeline**: 
![image](https://github.com/user-attachments/assets/61940b6c-9615-44d6-92e2-ec0619749e4f)
- `straightforward Lale Paper`: https://arxiv.org/pdf/2007.01977
![image](https://github.com/user-attachments/assets/9f6ff46b-7b4b-4210-89e0-9677de5a4d62)

---

1.5 **A Perspective on the Evolution of Live Programming** 
- https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=6617346
- `Liveness` as an attribute of a programming environment seems to have first been studied in the context of visual languages.
- Visual languages and program visualization attempt to solve a similar problem to the one that liveness addresses, which is the problem of **making programming easier by making it easier to understand quickly what a program is doing or supposed to do**. 

---

1.6 **Program Synthesis using Natural Language**

- 2016 , https://dl.acm.org/doi/pdf/10.1145/2884781.2884786?casa_token=Xy01MwJFlKUAAAAA:9_lT1jxIKUxRe51-RUbTflI7vp1JeAvBW_XdzD291B7Hbn2KtjnahXJKYSOivd8LYzNUmlLN1Fc

- early natural language-based program synthesis targets domain-specific language 
- unlike this paper, AI for low-code for AI paper used language model to deal with programming by natural language(PBNL)

---

**Comment by xp1632** [comment]
### 1.7 Low-Code Programming Models 

- https://dl.acm.org/doi/pdf/10.1145/3587691
- This paper is quite important! 
- it shows how we should define the view for the natural language component and visual programming component
- and to represent the program in **Domain-specific language (DSL )**
- In this case `Lale`, which is compatible with sklearn, Python 
![image](https://github.com/user-attachments/assets/03d5448d-0e78-47a1-9508-79b49054019d)
![image](https://github.com/user-attachments/assets/3693a0ce-aa58-454c-aefe-563e72008f13)

---

### 1.8 CodeT5: natural language <-> code generation model and used dataset
![image](https://github.com/user-attachments/assets/19fe8243-b7b5-4c1c-8647-110a1e2dba0e)
- https://arxiv.org/pdf/2109.00859
- CodeT5 was pretrained on corpus [CodeSearchNet](https://arxiv.org/pdf/1909.09436)
- CodeT5 was then fine-tuned on various tasks such as in the following table  from benchmark `CodeXGLUE` https://arxiv.org/pdf/2102.04664
![image](https://github.com/user-attachments/assets/b4156ff7-bd0d-47bf-afac-c6b3d37eb7e0)
